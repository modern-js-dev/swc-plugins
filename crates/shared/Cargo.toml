[package]
name = "shared"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
anyhow = "1.0.62"
serde_json = "1.0.71"
serde = {version= "1.0.71", features=["derive"]}
testing = "0.31.3"
swc_core = { features = [
  "common_concurrent",
  "base_concurrent",
  "ecma_ast",
  "ecma_visit",
  "ecma_loader_node",
  "ecma_loader_lru",
  "ecma_utils",
  "ecma_minifier",
  "ecma_transforms",
  "__ecma_transforms",
  "ecma_transforms_react",
  "ecma_transforms_typescript",
  "ecma_transforms_optimization",
  "ecma_quote",
  "ecma_parser",
  "ecma_parser_typescript",
  "cached",
  "base",
  "testing_transform"
], version = "0.40.20" }
dashmap = "5.4.0"
ahash = "0.7.6"

[dev-dependencies]
testing_macros = "0.2.7"
