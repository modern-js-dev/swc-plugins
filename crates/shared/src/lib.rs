pub mod utils;
use std::sync::Arc;

// reexports some same version libs
pub use anyhow;
pub use serde_json;
pub use swc_core;
use swc_core::common::{comments::SingleThreadedComments, Mark, SourceMap};
pub use swc_core::ecma::transforms::testing as swc_ecma_transforms_testing;
pub use testing;
pub extern crate serde;
pub use ahash;

pub struct PluginContext {
  pub cm: Arc<SourceMap>,
  pub top_level_mark: Mark,
  pub unresolved_mark: Mark,
  pub comments: SingleThreadedComments,
}
