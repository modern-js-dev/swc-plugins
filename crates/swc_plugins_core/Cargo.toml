[package]
edition = "2021"
name = "swc_plugins_core"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
plugin_import = {path = "../plugin_import"}
plugin_lock_corejs_version = {path = "../plugin_lock_corejs_version"}
plugin_lodash = {path = "../plugin_lodash"}
plugin_modernjs_ssr_loader_id = {path = "../plugin_modernjs_ssr_loader_id"}
plugin_modularize_imports = {path = "../plugin_modularize_imports"}
plugin_react_utils = {path = "../plugin_react_utils"}
plugin_remove_es_module_mark = {path = "../plugin_remove_es_module_mark"}
shared = {path = "../shared"}
swc_plugins_utils = {path = "../swc_plugins_utils"}

styled_components = "0.52.4"
styled_jsx = "0.29.4"
swc_emotion = "0.28.2"

[dev-dependencies]
insta = "1.18.2"
