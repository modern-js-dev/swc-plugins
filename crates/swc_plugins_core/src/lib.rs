mod minify;
mod pass;
mod transform;

pub mod types;
pub use minify::minify;
pub use transform::transform;

// plugins
pub use plugin_import;
pub use plugin_lock_corejs_version;
pub use plugin_lodash;
pub use plugin_modularize_imports;
pub use plugin_react_utils;
pub use styled_components as plugin_styled_components;
pub use styled_jsx as plugin_styled_jsx;
pub use swc_emotion as plugin_emotion;
