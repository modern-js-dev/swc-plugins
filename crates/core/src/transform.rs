use std::{path::PathBuf, sync::Arc};

use shared::{
  anyhow::Result,
  swc_core::{
    base::{config, try_with_handler, Compiler, HandlerOpts, TransformOutput},
    common::{errors::ColorConfig, FileName, Mark, GLOBALS},
    ecma::transforms::base::pass::noop,
  },
};

use crate::pass::internal_transform_pass;
use crate::types::TransformConfig;

pub fn transform(
  compiler: Arc<Compiler>,
  config: &TransformConfig,
  filename: String,
  code: &str,
  input_source_map: Option<String>,
) -> Result<TransformOutput> {
  GLOBALS.set(&Default::default(), || {
    let cm = compiler.cm.clone();

    try_with_handler(
      cm.clone(),
      HandlerOpts {
        color: ColorConfig::Never,
        skip_filename: false,
      },
      |handler| {
        compiler.run_transform(handler, true, || {
          let fm = cm.new_source_file(FileName::Real(PathBuf::from(&filename)), code.to_string());

          let mut swc_config = config::Options {
            ..config.swc.clone()
          };
          swc_config.config.input_source_map = input_source_map.map(config::InputSourceMap::Str);
          swc_config.filename = filename;

          let top_level_mark = swc_config.top_level_mark.unwrap_or_else(Mark::new);
          let unresolved_mark = Mark::new();

          swc_config.top_level_mark = Some(top_level_mark);

          compiler.process_js_with_custom_pass(
            fm,
            None,
            handler,
            &swc_config,
            // TODO pass comments to internal pass
            |_, _comments| internal_transform_pass(config, cm, top_level_mark, unresolved_mark),
            |_, _| noop(),
          )
        })
      },
    )
  })
}
