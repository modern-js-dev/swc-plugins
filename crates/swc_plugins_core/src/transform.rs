use std::{path::PathBuf, sync::Arc};

use shared::{
  anyhow::Result,
  serde_json,
  swc_core::{
    base::{
      config::{self, ModuleConfig},
      try_with_handler, Compiler, HandlerOpts, TransformOutput,
    },
    common::{errors::ColorConfig, FileName, Mark, GLOBALS, comments::SingleThreadedComments},
    ecma::{
      ast::EsVersion,
      parser::{Syntax, TsConfig},
      // TODO current version too low
      // transforms::module::common_js::Config
    },
  },
  PluginContext,
};
use swc_plugins_utils::is_esm;

use crate::pass::{internal_transform_after_pass, internal_transform_before_pass};
use crate::types::TransformConfig;

/// As we initialize plugins at each transform,
/// Some plugins need very heavy work on the first
/// time, and if we can cache it, we should get better
/// performance.
/// A `config_hash` is the unique key representing a
/// specific `TransformConfig`.
/// transform don't care how you create this hash.
/// 
/// If you call `transform` from `nodejs`, this config hash
/// is unique for each `binding::Compiler`.
pub fn transform(
  compiler: Arc<Compiler>,
  config: &TransformConfig,
  filename: String,
  code: &str,
  input_source_map: Option<String>,
  config_hash: Option<String>
) -> Result<TransformOutput> {
  GLOBALS.set(&Default::default(), || {
    let cm = compiler.cm.clone();

    try_with_handler(
      cm.clone(),
      HandlerOpts {
        color: ColorConfig::Never,
        skip_filename: false,
      },
      |handler| {
        compiler.run_transform(handler, true, || {
          let fm = cm.new_source_file(FileName::Real(PathBuf::from(&filename)), code.to_string());

          let mut swc_config = config::Options {
            ..config.swc.clone()
          };
          swc_config.config.input_source_map = input_source_map.map(config::InputSourceMap::Str);
          swc_config.filename = filename;

          let top_level_mark = swc_config.top_level_mark.unwrap_or_else(Mark::new);
          let unresolved_mark = Mark::new();

          swc_config.top_level_mark = Some(top_level_mark);
          let comments = SingleThreadedComments::default();

          // Need auto detect esm
          let program = if swc_config.config.module.is_none() {
            let program = compiler.parse_js(
              fm.clone(),
              handler,
              swc_config.config.jsc.target.unwrap_or(EsVersion::Es2022),
              swc_config.config.jsc.syntax.unwrap_or_else(|| {
                Syntax::Typescript(TsConfig {
                  tsx: true,
                  decorators: true,
                  ..Default::default()
                })
              }),
              config::IsModule::Bool(true),
              Some(&comments),
            )?;

            swc_config.config.module = Some(if is_esm(&program) {
              ModuleConfig::Es6
            } else {
              ModuleConfig::CommonJs(
                serde_json::from_str(
                  r#"{
                    "ignoreDynamic": true
                  }"#,
                )
                .unwrap(),
              )
            });
            Some(program)
          } else {
            None
          };

          // TODO comments can be pass to `process_js_with_custom_pass` in next swc version
          let plugin_context = Arc::new(PluginContext {
            cm,
            top_level_mark,
            unresolved_mark,
            comments: comments.clone(),
            config_hash,
          });

          compiler.process_js_with_custom_pass(
            fm,
            program,
            handler,
            &swc_config,
            comments,
            // TODO pass comments to internal pass in next swc versions
            |_| {
              internal_transform_before_pass(
                config,
                plugin_context.clone()
              )
            },
            |_| {
              internal_transform_after_pass(
                config,
                plugin_context.clone()
              )
            },
          )
        })
      },
    )
  })
}
