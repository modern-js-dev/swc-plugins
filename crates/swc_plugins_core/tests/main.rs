use swc_plugins_core::transform;
use shared::swc_core::{
  base::{config::Options, Compiler},
  common::SourceMap
};
use std::sync::Arc;

#[test]
fn test() {
  let code = "const a = {};
  const b = { ...a }";

  let config = swc_plugins_core::types::TransformConfig {
    swc: Options {
      config: shared::serde_json::from_str(
        r#"{
        "jsc": {
          "externalHelpers": true
        }
      }"#,
      )
      .unwrap(),
      ..Default::default()
    },
    extensions: Default::default(),
  };

  let res = transform(
    Arc::new(Compiler::new(Arc::new(SourceMap::default()))),
    &config,
    "".into(),
    code,
    None,
  )
  .unwrap();
  insta::assert_snapshot!("plugin-import", res.code);
}
