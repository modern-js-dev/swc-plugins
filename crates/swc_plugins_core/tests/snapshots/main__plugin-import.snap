---
source: crates/swc_plugins_core/tests/main.rs
assertion_line: 36
expression: res.code
---
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _objectSpread = require("@swc/helpers/lib/_object_spread.js").default;
var a = {};
var b = _objectSpread({}, a);

